Instructions improvements (#11)
input-output-hk/dappstore-cip72: README.md

## What is on-chain DApp registration?

The on-chain DApp registration, based on CIP-72 (Cardano DApp registration standard), is a transaction with specific
metadata describing your Dapp. In general, the on-chain DApp registration process is similar to the on-chain Stake Pool
registration process, but in this case the DApp developer references key information about their DApp (e.g. DApp name,
logo, description, contacts, website, etc.) stored somewhere off-chain (e.g. DApp website, Github, etc.) on Cardano blockchain.

#### We can see the following benefits for the DApp developer from registering their DApp on Cardano blockchain:
1. remove the hassle of providing the same key information about your DApp to various sources each time you are asked to
2. advertise and anchor your DApp on Cardano blockchain
3. make your DApp easily discoverable from day one by Cardano community 
4. make key information about your DApp available in a standardized way so any DApp aggregator, board, or wallet can
integrate and consume it easily

## Create Off-chain JSON with DApp data

Initiate your dApp's on-chain registration journey by crafting an offchain-metadata.json file. 

> Seek inspiration from the 16 sample offchain-metadata.json files conveniently housed within the [examples](tree/main/examples) folder. 

Author: ```suggestion
Seek inspiration from the 16 sample offchain-metadata.json files conveniently housed within the [examples](./examples) folder. 
```

# How to register dApp to Cardano

Dapp registration is a transaction with specific metadata describing your dapp.

> For more details refer to [cip-72 specification](https://developers.cardano.org/docs/governance/cardano-improvement-proposals/cip-0072/).

Reviewer: Should we point to cardano.org or GitHub, since GitHub is often more up-to-date?

This repository consists of scripts that intend to help you with submitting the registration transaction.

## Set up required tooling

You will need a node.js and yarn installed in your system.

You need a few tools to be available in your terminal:
 - `cardani-cli`
 - `cardano-address`

>  - `bech32`

Reviewer: Shouldn't we provide them with info on how to verify that these are installed? (e.g verify that it is installed using `bech32 -v`). Also, can suggest they use `cabal` to install these since it's an easier approach and would save the developer time.

Probably at the bottom of the ReadMe, we can add something like a glossary section that contains more info about installing these three packages.

Author: A few lines below there are two recommended ways to get the required tools (cardano-wallet or Daedalus). Probably you could get them in plenty of other ways but I don't think we should list more ways. Many of developers are haskelers so it will be obvious for them to use `cbal`

IMO we also should not add a note for them to test whether those tools work. Isn't that obvious to run the tool to test it? 😉 In the section recommending cardano-wallet or Daedalus it is described to set the $PATH env variable properly, so either all of them will work on none.

You can set up a local cardano node if you choose to submit the registration transaction this way.
Alternatively you could use [https://blockfrost.io/](https://blockfrost.io/). To do so you need to get an api key.

We recommend the following ways of getting required binaries and the local node:

 - [Latests cardano-wallet release](https://github.com/cardano-foundation/cardano-wallet/releases/latest): Downlaod the `cardano-wallet` archive appropriate for your system, unpack it
and put the unpacked directory in you $PATH variable so the contents are available in your terminal. Start the local
node if you which to use it.

>  - [Daedalus](https://daedaluswallet.io/en/download/): Download the `Daedalus` for chosen network, install it and add

Reviewer: Daedalus download page only has the package for mainnet. User should be able to download daedalus for any network.

```suggestion
 - [Daedalus](https://buildkite.com/input-output-hk/daedalus/builds/22815): Download the `Daedalus` for chosen network, install it and add
```

the `location-of-instalation/bin` to the $PATH variable so the contents are available in your terminal. This is an easy
way to spin up a node locally however you will need to get the `bech32` yourself because it is not included in the
Daedalus files

### Testnet-magic numbers 

In few places you will have to provide proper network parameter `(--mainnet | --testnet-magic NATURAL)`

- 2: Preview. `--testnet-magic=2`
- 1: Preprod. `--testnet-magic=1`

### Cardano node socket port

If you choose to go with the local node you need to have the `CARDANO_NODE_SOCKET_PORT` env variable set to the path of
the local node socket file.

## Step 1: Set up wallet

There are three ways to setup a wallet.

- a) Restore wallet using mnemonics.
- b) Restore a wallet using existing keys (without mnemonics).
- c) Create new wallet

### a) Restore wallet from mnemonic

1. Create `phrase.prv` file and fill it up with your mnemonic in space seperated format.
2. Grant script permission to run `chmod +x ./scripts/restore-mnemonic-wallet.sh`
3. Restore mnemonic wallet by running `NETWORK=(preview|preprod|mainnet) LOCAL_NODE=(true|false) yarn restore-mnemonic-wallet`

> (Choose appropriate values of the NETWORK and LOCAL_NODE variables)

Reviewer: ```suggestion
(Choose appropriate values of the NETWORK and LOCAL_NODE variables)
4. `cardano-cli query tip (--mainnet | --testnet-magic NATURAL)` to verify that the local node is properly setup
```

Author: This command is applicable only if the user chooses to use the local node instead of blockfrost. Moreover, the previous script will fail on querying the protocol parameter if something is not working correctly so I don't see a big value in this step TBH.

### b) Restore a wallet using existing keys

If you have your `payment.skey`, `payment.vkey`, `stake.skey` and `stake.vkey`, then you can follow these steps

1. Create a payment address
`$ cardano-cli address build --payment-verification-key-file payment.vkey --stake-verification-key-file stake.vkey --out-file payment.addr`
2. (LOCAL NODE ONLY) Retrieve the node’s current pool parameters
`$ cardano-cli query protocol-parameters --out-file protocol.json (--mainnet | --testnet-magic NATURAL)`
(Choose appropriate network parameter)

### c) Create a new wallet

1. Grant script permission to run `chmod +x ./scripts/setup-new-wallet.sh`
2. Setup new wallet. `NETWORK=(preview|preprod|mainnet) LOCAL_NODE=(true|false) yarn setup-new-wallet`
(Choose appropriate values of the NETWORK and LOCAL_NODE variables)
3. Request Test ADA at [https://docs.cardano.org/cardano-testnet/tools/faucet](https://docs.cardano.org/cardano-testnet/tools/faucet)
to the generated address from the `payment.addr` file. NB transaction fee it's a variable amount, something around 190.000 lovelace

## Step 2: Run the registration script

> 1. Make sure your wallet has test ADA for transaction fee

Reviewer: Should we add instructions on how to verify that the wallet has transaction fees?

```suggestion
1. Make sure your wallet has test ADA for transaction fee. Check the useful command section below on a command to query the utxo ).
```

2. Install packages `yarn install`
3. Copy `.env.example-blockfrost` or `.env.example-local-node` to the `.env` file and edit it appropriately
`cp .env.example-<blockfrost|local-node> .env`

> 4. Create a cip72 metadata file.

Reviewer: ```suggestion
4. Create and host a cip72 metadata file (we recommend you use GitHub gist).
```

Author: Considering recent Volo's changes which ask the user to create it at the beginning I just mentioned to prepare the link created before
